File: 316.remove-duplicate-letters.py
#
# @lc app=leetcode id=316 lang=python3
#
# [316] Remove Duplicate Letters
#

# Time:  O(n)
# Space: O(1)

from collections import defaultdict


# @lc code=start
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        res, count, visited = ["0"], defaultdict(int), defaultdict(bool)
        for c in s:
            count[c] += 1
        for c in s:
            count[c] -= 1
            if visited[c]:
                continue
            while c < res[-1] and count[res[-1]]:
                visited[res[-1]] = 0
                res.pop()
            res.append(c)
            visited[c] = True
        return "".join(res[1:])


# @lc code=end

File: 316.remove-duplicate-letters.js
/*
 * @lc app=leetcode id=316 lang=javascript
 *
 * [316] Remove Duplicate Letters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var removeDuplicateLetters = function (s) {
  const res = ["0"];
  const count = {};
  const visited = {};
  for (let i = 0; i < s.length; ++i) {
    count[s[i]] = count[s[i]] + 1 || 1;
  }
  for (let i = 0; i < s.length; ++i) {
    --count[s[i]];
    if (visited[s[i]]) {
      continue;
    }
    while (s[i] < res[res.length - 1] && count[res[res.length - 1]]) {
      visited[res[res.length - 1]] = 0;
      res.pop();
    }
    res.push(s[i]);
    visited[s[i]] = true;
  }
  return res.slice(1).join("");
};
// @lc code=end
